Add timeout for waiting on compactor to become ACTIVE in the ring.

ac1214 opened the pull request:
Signed-off-by: Albert <ac1214@users.noreply.github.com>

**What this PR does**:
Addresses #3914
**Which issue(s) this PR fixes**:
Fixes #3914

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

bboreham commented on pkg/compactor/compactor.go:
```diff
@@ -41,8 +41,9 @@ const (
-	RingOp                = ring.NewOp([]ring.InstanceState{ring.ACTIVE}, nil)
+	errInvalidBlockRanges            = "compactor block range periods should be divisible by the previous one, but %s is not divisible by %s"
+	RingOp                           = ring.NewOp([]ring.InstanceState{ring.ACTIVE}, nil)
+	WaitInstanceStateTimeoutDuration = time.Duration(600)
```
This seems weird. `time.Duration(600)` is 600ns.
I recommend you make this `600*time.Second` (or perhaps `10*time.Minute`) and remove the multiplication on line 398.

Also why isn't it a constant?  (Maybe it should be configurable, but you didn't do that)

ac1214 replied:
Thanks for the review, I made changes to allow the timeout to be configurable.

bboreham commented on pkg/ring/ring.go:
```diff
@@ -763,8 +767,7 @@ func (r *Ring) shuffleShard(identifier string, size int, lookbackPeriod time.Dur
 
-// GetInstanceState returns the current state of an instance or an error if the
-// instance does not exist in the ring.
+// GetInstanceState implements ReadRing.
```
It would be ok to leave the description here.

bboreham reviewed (commented):
Thanks for the PR! It's broadly ok, but a couple of style points.

It looks to me like the tests you've added test similar code, rather than the code you added. If I'm right, could you test the actual code?

What is the user experience if the timeout is hit?  Do they see anything in the logs to let them know what happened?

ac1214 commented:
> Thanks for the PR! It's broadly ok, but a couple of style points.
> 
> It looks to me like the tests you've added test similar code, rather than the code you added. If I'm right, could you test the actual code?
> 
> What is the user experience if the timeout is hit? Do they see anything in the logs to let them know what happened?

I added tests to actually test the starting of the compactor. I also added some logs as to make it clear to why the compactor failed to start.

bboreham commented:
Sorry but we have begun the process of cutting a new release; please rebase from `master` and move your CHANGELOG entry to the top under `## master / unreleased`

pracucci commented on pkg/compactor/compactor.go:
```diff
@@ -393,7 +393,11 @@ func (c *Compactor) starting(ctx context.Context) error {
 		// to this shard or not).
 		level.Info(c.logger).Log("msg", "waiting until compactor is ACTIVE in the ring")
+
+		ctx, cancel := context.WithTimeout(ctx, c.compactorCfg.ShardingRing.StartingTimeout)
```
We shouldn't overwrite `ctx` otherwise the timeout applies to next steps too. You should name the returned context variable differently and use it below where you want the timeout to be applied.

pracucci commented on pkg/compactor/compactor_ring.go:
```diff
@@ -59,6 +61,9 @@ func (cfg *RingConfig) RegisterFlags(f *flag.FlagSet) {
 	f.StringVar(&cfg.InstanceID, "compactor.ring.instance-id", hostname, "Instance ID to register in the ring.")
+
+	// Timeout durations
+	f.DurationVar(&cfg.StartingTimeout, "compactor.ring.starting-timeout", 10*time.Minute, "Timeout for waiting on compactor to become ACTIVE in the ring")
```
[nit] Period at the end.

```suggestion
	f.DurationVar(&cfg.StartingTimeout, "compactor.ring.starting-timeout", 10*time.Minute, "Timeout for waiting on compactor to become ACTIVE in the ring.")
```

pracucci replied:
Sorry for being picky. I would make the CLI flag (and the YAML config option) name even more explicit:
- CLI flag: `compactor.ring.wait-active-instance-timeout`
- YAML field: `wait_active_instance_timeout`

pracucci commented on pkg/compactor/compactor_test.go:
```diff
@@ -1279,3 +1282,29 @@ func TestCompactor_DeleteLocalSyncFiles(t *testing.T) {
+	// Try to start the compactor with a bad consul kv-store. The
+	err := services.StartAndAwaitRunning(context.Background(), c)
+
+	// Assert that the compactor timesout
```
[nit]
```suggestion
	// Assert that the compactor timed out
```

pracucci commented on pkg/ring/util_test.go:
```diff
@@ -184,3 +239,63 @@ func TestWaitRingStabilityShouldReturnErrorIfMaxWaitingIsReached(t *testing.T) {
+
+	const (
+		instanceID      = "test"
+		timeoutDuration = time.Duration(3)
```
I guess you wanted 3s and not 3ns. You should change this to `time.Duration(3 * time.Second)`. I think 1s is enough for this test, so I would just do `time.Second`.

_Same comment applies to other tests._

pracucci reviewed (commented):
Good job and thanks for addressing my feedback. Way easier than the initial version. I left few last comments 🙏

pracucci commented on CHANGELOG.md:
```diff
@@ -3,6 +3,7 @@
 ## master / unreleased
 
 * [CHANGE] Querier / ruler: Change `-querier.max-fetched-chunks-per-query` configuration to limit to maximum number of chunks that can be fetched in a single query. The number of chunks fetched by ingesters AND long-term storare combined should not exceed the value configured on `-querier.max-fetched-chunks-per-query`. #4260
+* [CHANGE] Add timeout for waiting on compactor to become ACTIVE in the ring. #4262
```
[nit] We use CHANGE for breaking changes. This is more an enhancement.

```suggestion
* [ENHANCEMENT] Add timeout for waiting on compactor to become ACTIVE in the ring. #4262
```

ac1214 replied:
Fixed! Thanks for the review!

pracucci reviewed (approved):
LGTM (modulo a nit)

ac1214 commented:
Would I be able to get a second review on this PR and get the workflow triggered again? @bboreham
Thanks!

bboreham reviewed (approved):
LGTM
